extensions/objects/templates/image-editor.component.ts

```diff
@@ -0,0 +1,1155 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Component for filepath editor.
+ */
+
+// This component can only be used in the context of an exploration.
+/**
+ * NOTE: These are some hacky behavior in the component class. Apologies in
+ * advance if you were stuck on one of these for a long time.
+ *
+ * 1. Whenever changing a property of the data property of this component
+ * class, please make sure the object is reassigned. Angular's change detection
+ * uses the `===` operator which would only return false when the object
+ * changes. You can use the spread operator to achieve the desired result.
+ * Eg: this.data = {..this.data};
+ *
+ * 2. Angular treats SVG+XMl as dangerous by default. To show this image in the
+ * view, we run our own security (for our own safety) and circumvent the angular
+ * security. When we circumvent Angular's security checks, it attaches some
+ * additional information to inform the Angular's view engine to not run
+ * security checks on the image data. This can only be done the in html file
+ * using property binding (i.e. [src]). If we use this in the ts file, by doing
+ * `const img = new Image; img.src = circumventedValue;`, Angular will raise
+ * errors.
+ * In addition to this, our custom functions will also fail for this data. To
+ * overcome this, during the migration of this component, a new property called
+ * imgData was introduced. This contains the not circumvented value for svgs or
+ * just the normal image data for other file formats. While updating
+ * `this.data.metadata.uploadedImageData`, make sure that you also update
+ * imgData. Failing to do so could lead to unpredictable behavior.
+ */
+import { Component, EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges } from '@angular/core';
+import { SafeResourceUrl } from '@angular/platform-browser';
+import { downgradeComponent } from '@angular/upgrade/static';
+import { AppConstants } from 'app.constants';
+import { UrlInterpolationService } from 'domain/utilities/url-interpolation.service';
+import { ImagePreloaderService } from 'pages/exploration-player-page/services/image-preloader.service';
+import { AlertsService } from 'services/alerts.service';
+import { AssetsBackendApiService } from 'services/assets-backend-api.service';
+import { ContextService } from 'services/context.service';
+import { CsrfTokenService } from 'services/csrf-token.service';
+import { ImageLocalStorageService } from 'services/image-local-storage.service';
+import { ImageUploadHelperService } from 'services/image-upload-helper.service';
+import { SvgSanitizerService } from 'services/svg-sanitizer.service';
+
+const gifFrames = require('gif-frames');
+const gifshot = require('gifshot');
+
+interface FilepathData {
+  mode: number;
+  metadata: {
+    uploadedFile?: File;
+    uploadedImageData?: string | SafeResourceUrl;
+    originalWidth?: number;
+    originalHeight?: number;
+    savedImageFilename?: string;
+    savedImageUrl?: string;
```

Comment:
Not entirely sure.. I basically moved stuff from filepath-editor to this file as-is.

I tried removing `?` and ran the typescript checks, and it seems like we initialise metadata to `{}` in line 217. What should be done in this case?

Reply:
@EeshaArif @mridul-netizen Can you help here?

Reply:
I think this makes sense as long as we're passing different set of properties everywhere for metdata. But a better approach would be to not make them optional so that structure of medata remains same everywhere. Here optional properties can be separated out to some interface like:
```typescript
interface MetaData {
  uploadedFile: File;
  uploadedImageData: string | SafeResourceUrl;
  originalWidth: number;
  originalHeight: number;
  savedImageFilename: string;
  savedImageUrl: string;
}
. . .
interface FilepathData {
  mode: number;
  metadata: MetaData;
  crop: boolean;
}
```
and then we can typecast it as:
```typescript
this.data = { mode: this.MODE_EMPTY, metadata: <MetaData>{}, crop: true };
```
Ditto elsewhere.
@kevintab95 PTAL!

Reply:
we could also directly typecast metadata like this
```typescript
this.data = { mode: this.MODE_EMPTY, metadata: <FilepathData['metadata']>{}, crop: true };

```
but mridul's approach of separating into interfaces seems more cleaner